### ISSUE TITLE
Set image tag preserving image name

### ISSUE THREAD

**Issue by alecunsolo** [kubernetes-sigs/kustomize#3487]
When using `kustomize edit set image` to change the tag of an image the `newName` is not preserved
I should be able to change only the image tag and keep the `newName` as-is

This is similar to #2625, but the proposed solution doesn't work as I expected.
This is the source kustomization
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

images:
- name: myname
  newName: newName
  tag: v1.0.0
```
After `kustomize edit set image myname=:v2.0.0` the resulting kustomization.yaml is
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

images:
- name: myname
  newTag: v2.0.0
```

I'm using version 3.9.2:
```sh
kustomize version
{Version:kustomize/v3.9.2 GitCommit: BuildDate:2021-01-17T19:01:50+00:00 GoOs:darwin GoArch:amd64}
```

**Comment by Shell32-Natsu** [comment]
You are right it's not supported now. I thought I misunderstood the context in #2625. We may need to have a new subcommand to modify newTag/newName individually.

**Comment by dimw** [comment]
**tl;dr** 

Proposal for extension of the `set image` command for keeping `newName`/ `newTag`:
```
$ kustomize edit set image image1=<preserve-indicator>:v1
```

**Detailed Background**

I am facing [the same need](https://github.com/kubernetes-sigs/kustomize/issues/3487#issue-789894507) for keeping `newName` when setting the `newTag`. In my use-case I have multiple resource overlays (i.e. _dev_ and _prod_) which use images from different registries. Thus, predefining the different `newName`s in the overlay kustomization files sounds handy. However, similar to @alecunsolo, I need the possibility to rewrite the `newTag` only programmatically from the CI.

Introduction of a [new sub-command](https://github.com/kubernetes-sigs/kustomize/issues/3487#issuecomment-764959551) to modify the `newTag`/`newName` proposed by @Shell32-Natsu sounds like a good idea. However, in an environment where multiple images have to be set (see examples of `$ kustomize edit set image --help`) this might result in the necessity for executing multiple commands, e.g.:
```
$ kustomize edit set image-tag image1=v1
$ kustomize edit set image-name image2=my-image2
$ kustomize edit set image image3=my-image3:v3
```

To avoid the situation depicted above I would propose to implement a backward compatible solution which is working with the existing `set image` command while preserving `newName`/`newTag`. Particularly, the constrains for the image name/tag according to [Docker documentation](https://docs.docker.com/engine/reference/commandline/tag/#extended-description) are the following:
> An image name is made up of slash-separated name components, optionally prefixed by a registry hostname. The hostname must comply with standard DNS rules, but may not contain underscores. [...] Name components may contain lowercase letters, digits and separators. A separator is defined as a period, one or two underscores, or one or more dashes. A name component may not start or end with a separator.
>
> A tag name must be valid ASCII and may contain lowercase and uppercase letters, digits, underscores, periods and dashes. [...]

Thus, I would suggest using a "magic" string/character which is forbidden either for usage in image name or in image tag as an indicator for Kustomize for keeping the `newName`/`newTag` currently defined, the so called _preserve-indicator_.  Example usage pattern:
```
$ kustomize edit set image image1=<preserve-indicator>:v1 image2=my-image2:<preserve-indicator>
```

It looks like using the character `*` (asterisk) which is also used as wildcard symbol and thus stands for something which is already there could be an intuitive solution. However, this would end up in the necessity for quoting the flags (which is btw. usually a good idea) to avoid interference with shell.

An example command for setting the tag while preserving the name would look as following:
```
$ kustomize edit set image "image1=*:v1"
```
While setting the name while keeping the tag would result in:
```
$ kustomize edit set image "image2=my-image2:*"
```
Of course, both can be combined into one command:
```
$ kustomize edit set image "image1=*:v1" "image2=my-image2:*"
```

It is up for discussion whether another string such as `.` (dot), `/` (slash), `$` (dollar), or `{}` should be used as _preserve-indicator_. Again, please keep in mind that the string should be ideally an invalid image name/tag and should not produce any side-effects in shell. Of course, we can also introduce a dedicated `--preserve-indicator` flag and pass the string alongside with the command.

**Comment by alecunsolo** [comment]
I like the 'magic character' approach. As poinded by @dimw it should be backward compatible and shorter than the subcommand.
It's not the mentioned use case, but as a generic feature it should support `digest`s too:
```sh
kustomize edit set image 'image1=*@<digest>'
```

**Comment by dimw** [comment]
@alecunsolo As suggested, I also added your digest preservation use-case. Further, because it was not clear to me, I also run some tests and confirmed that it is possible to have different image names with the same digest (also see [Docker image digest documentation](https://docs.docker.com/engine/reference/commandline/images/#list-image-digests)). Thus, I am also considering the inverse use-case now. 

Example usage:
```
$ kustomize edit set image "image1=*@<digest>"
$ kustomize edit set image "image1=<image-name>@*"
```

**Comment by Shell32-Natsu** [comment]
Close this in #3503
